lib/auth/userpreferences/userpreferencesv1/service_test.go

```diff
@@ -0,0 +1,232 @@
+		},
+		Theme: userpreferencesv1.Theme_THEME_LIGHT,
+		Onboard: &userpreferencesv1.OnboardUserPreferences{
+			PreferredResources: []userpreferencesv1.Resource{},
+		},
+	}
+
+	tests := []struct {
+		name     string
+		userName string
+		req      *userpreferencesv1.UpsertUserPreferencesRequest
+		wantErr  assert.ErrorAssertionFunc
+	}{
+		{
+			name:     "success",
+			userName: defaultUser,
+			req: &userpreferencesv1.UpsertUserPreferencesRequest{
+				Preferences: defaultPreferences,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name:     "access denied - user doesn't exist",
+			userName: nonExistingUser,
+			req: &userpreferencesv1.UpsertUserPreferencesRequest{
+				Preferences: defaultPreferences,
+			},
+			wantErr: assert.Error,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctxs, svc := initSvc(t)
+
+			_, err := svc.UpsertUserPreferences(ctxs[tt.userName], tt.req)
+			if tt.wantErr(t, err) {
+				return
+			}
+		})
+	}
+}
+
+func initSvc(t *testing.T) (map[string]context.Context, *Service) {
+	ctx := context.Background()
+	backend, err := memory.New(memory.Config{})
+	require.NoError(t, err)
+
+	clusterConfigSvc, err := local.NewClusterConfigurationService(backend)
+	require.NoError(t, err)
+	trustSvc := local.NewCAService(backend)
+	roleSvc := local.NewAccessService(backend)
+	userSvc := local.NewIdentityService(backend)
+
+	require.NoError(t, clusterConfigSvc.SetAuthPreference(ctx, types.DefaultAuthPreference()))
+	require.NoError(t, clusterConfigSvc.SetClusterAuditConfig(ctx, types.DefaultClusterAuditConfig()))
+	require.NoError(t, clusterConfigSvc.SetClusterNetworkingConfig(ctx, types.DefaultClusterNetworkingConfig()))
+	require.NoError(t, clusterConfigSvc.SetSessionRecordingConfig(ctx, types.DefaultSessionRecordingConfig()))
+
+	accessPoint := struct {
+		services.ClusterConfiguration
+		services.Trust
+		services.RoleGetter
+		services.UserGetter
+	}{
+		ClusterConfiguration: clusterConfigSvc,
+		Trust:                trustSvc,
+		RoleGetter:           roleSvc,
+		UserGetter:           userSvc,
+	}
+
+	accessService := local.NewAccessService(backend)
+	eventService := local.NewEventsService(backend)
+	lockWatcher, err := services.NewLockWatcher(ctx, services.LockWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Client:    eventService,
+			Component: "test",
+		},
+		LockGetter: accessService,
+	})
+	require.NoError(t, err)
+
+	authorizer, err := authz.NewAuthorizer(authz.AuthorizerOpts{
+		ClusterName: "test-cluster",
+		AccessPoint: accessPoint,
+		LockWatcher: lockWatcher,
+	})
+	require.NoError(t, err)
+
+	roles := map[string]types.Role{}
+
+	role, err := types.NewRole("allow-rules", types.RoleSpecV6{
+		Allow: types.RoleConditions{
+			Rules: []types.Rule{
+				{
+					Resources: []string{types.KindUser},
+					Verbs:     []string{types.VerbList, types.VerbRead, types.VerbUpdate, types.VerbCreate, types.VerbDelete},
+				},
+			},
+		},
+	})
```

Comment:
It should work without these permissions
```suggestion
	role, err := types.NewRole("allow-rules", types.RoleSpecV6{})
```

Reply:
Removed.